core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -1156,6 +1158,28 @@ private[spark] object Utils extends Logging {
   }
 
   /**
+   * Execute a block of code that evaluates to Unit, stop SparkContext is there is any uncaught 
+   * exception
+   *  
+   * NOTE: This method is to be called by the driver-side components to avoid stopping the 
+   * user-started JVM process completely; in contrast, tryOrExit is to be called in the 
+   * spark-started JVM process .
+   */
+  def tryOrStopSparkContext(sc: SparkContext)(block: => Unit) {
+    try {
+      block
+    } catch {
+      case e: ControlThrowable => throw e
+      case t: Throwable =>
+        if (sc != null) {
+          logError(s"uncaught error in thread ${Thread.currentThread().getName}, stopping " +
+            "SparkContext", t)
+          sc.stop()
+        }
```

Comment:
Hi, @zsxwing thanks for the comments

I personally prefer a more conservative way here (the current approach)

Because the throwable thrown from here can be varying in terms of types, and I'm concerning that the Throwable from here, like `OOM`, would be mixed with the instances of the same type generated by the other components in user's program; on the other hand, our goal is just to let the user know `SparkContext` is stopped

So I prefer to letting the user call SparkContext.runJob to get a 'new IllegalStateException("SparkContext has been shutdown")' which (hopefully) will be handled `exactly`

@srowen @aarondav , your thoughts?